Documentation/General/emergentArchitectureDesign/emergentArchitectureDesign.tex

```diff
@@ -0,0 +1,180 @@
+In these rooms the user can interact with other listeners through the chat while listening to the music.
+
+\begin{figure}[h!]
+\includegraphics[scale=.4]{useCases.png}
+\caption{Use case diagram for Moodcat}
+\label{fig:MoodCatUseCase}
+\end{figure}
+
+\par
+The MoodCat frontend will be web-based.
+This allows us to support all modern devices which enables a large target audience.
+We will develop the frontend using modern techniques as \Gls{HTML}\cite{HTML}, \Gls{JS} and \Gls{CSS}\cite{CSS}.
+Twitter Bootstrap will be used as boilerplate for our own customized CSS thema.
+
+\par
+AngularJS\cite{AngularJS} will be used as web framework, providing us two-way data binding between the user interface and backend, and a standardized way of composing user interface controlling logic through \textit{controllers}, \textit{services} and \textit{directives}.
+
+\par
+The Object-oriented\cite{Wirfs-Brock} and Dependency Injection\cite{angularDI} driven nature of AngularJS makes it suitable for unit testing\cite{Binder}.
+Testing the frontend codebase will be done using the Karma\cite{Karma} testrunner, the Jasmine\cite{Jasmine} testing framework and the PhantomJS\cite{PhantomJS} headless\footnote{JavaScript runs in browsers and the written logic needs a DOM to interact with. A headless browser does not actually render the page and is therefore much faster in intergration tests.} browser.
+
+\subsection{Moodcat Backend}
+The Moodcat backend has several subcomponents, each with their own responsibility:
+\begin{itemize}
+\item the \textit{Static File Server} that serves the frontend website on a webserver;
+\item the \textit{REST API} which is used for communication about rooms and songs between the backend and frontend;
+\item the \textit{Music matching algorithm} which is used to generate music suggestions for rooms and to find a room for a user.
+\end{itemize}
+In the following paragraphs we will explain the last two subcomponents.
+
+\subsubsection{REST API}
+The MoodCat REST\cite{rest} connects the frontend to the music matching algorithm.
+The backend keeps track of which rooms are active and what song is currently played in the room.
+Furthermore, the backend is responsible for transmitting the chat messages between the users in the room.
+An example interaction between the frontend and the backend is given in \ref{fig: backendInteraction}.
+% I am not so happy that we do not mention how we rely on Soundcloud for the streaming here, but I have the feeling it does not really fit here. Suggestions?
+
+\begin{figure}[h!]
+\includegraphics[scale=.4]{backendInteraction.png}
+\caption{Example interaction frontend and backend}
+\label{fig: backendInteraction}
+\end{figure}
+
+
+\par
+The backend and the frontend communicate over \Gls{HTTP} requests in the \Gls{JSON} format.
+% We might want to note Websockets at a later point
```

Comment:
I'd say include it now, mentioning "As an extension and improvement of the usability"